### ISSUE TITLE
Grails 3.0.10: messages.properties should be loaded in the same sequence as plugins are loaded.

### ISSUE THREAD

**Issue by satshil** [grails/grails-core#9502]
Environment
================
java 1.7.80
grails 3.0.10

Steps to reproduce
================
1. create grails app - testAPP
2. create plugin - testPlugin1
3 create plugin - testPlugin2 (depends n plugin1)
3. add dependency for plugin inside app
4. add some message in plugin2's messages.properties
5. override some message in plugin 1's messages.properties
4. run clean on app and plugin
5. run app

problem
================
If debugged and look into following code the plugin sequence is not maintained.
```
protected void mergeBinaryPluginProperties(final Locale locale, Properties mergedProps) {
        final GrailsPlugin[] allPlugins = pluginManager.getAllPlugins();
        for (GrailsPlugin plugin : allPlugins) {
            if (plugin instanceof BinaryGrailsPlugin) {
                BinaryGrailsPlugin binaryPlugin = (BinaryGrailsPlugin) plugin;
                final Properties binaryPluginProperties = binaryPlugin.getProperties(locale);
                if (binaryPluginProperties != null) {
                    mergedProps.putAll(binaryPluginProperties);
                }
            }
        }
    }
```

Reason for this change
========
having maintained the sequence we provide advantage to override the plugin's messages in another plugins.

**Comment by ctoestreich** [comment]
@satshil What if your application (a) uses a plugin (b) that adds plugin (c) but the developer doesn't know about (c) and adds the dependency to (a) so that  

a->b->c
a->c

What order would (c) message.properties be loaded?  How would you know the actual intent?

I think it is probably a better practice to let your application (a) override the properties that are needed to  be customized at runtime and not have plugins overriding other plugins properties as a way to control message property load order.  

We have also run into this, but we defer to the actual containing application (a) to decide the end state of non-generic message properties.

**Comment by satshil** [comment]
@ctoestreich,

Thanks for reply.

What I am looking for when I suggesting for this change, let me explain with your scenario
Scenario :  a [app]-> b[plugin] -> c[plugin]
1. Application will always remain super which can override any message from any plugin.
2. c is the plug in which should loaded first followed by b then a.
3. whats happening currently is we are creating the collection of messages and merging it which does not have any guaranteed sequence hence it might bring surprise in runtime environment. Hence good to go with well defined approach. Easy to debug and find error.

Same approach is implemented for URLMappings in #9441

**Comment by Schlogen** [comment]
@satshil If plugin b defines `loadAfter = ['c']` then the i18n messages are in the right order. I don't think it is possible to determine which plugins should be loaded in which order without that being specified.
